src/popsift/popsift.cpp

```diff
@@ -317,3 +316,30 @@ popsift::FeaturesDev* SiftJob::getDev()
     return dynamic_cast<popsift::FeaturesDev*>( _f.get() );
 }
 
+void PopSift::Pipe::uninit()
+{
```

Comment:
@griwodz I replicated the uninit inside Pipe, it's basically the same code as before plus those nullptr safe checks.
Just checking whether the order of operations need to be exactly the same:
before it was `join()` for 2 and 1 and then delete:
```c++ 
_pipe._thread_stage2->join();
_pipe._thread_stage1->join();
delete _pipe._thread_stage2;
delete _pipe._thread_stage1;
```
Here instead i delete after `join()` for each.
I don't know if it makes a difference (i.e. somehow there is shared data or something between the two threads).

Reply:
If I remember correctly, it is safe to join both before delete. Neither of the threads will terminate before the producer-consumer queues between them are drained, so all critical state should be gone when both joins have returned.

Reply:
Ok, but in the new version it is
```c++
    if(_thread_stage2 != nullptr)
    {
        _thread_stage2->join();
        delete _thread_stage2;
    }
    if(_thread_stage1 != nullptr)
    {
        _thread_stage1->join();
        delete _thread_stage1;
    }
```
i.e. is it ok if we delete before the other join()?

Reply:
Yes, the delete operation kills only the boost::thread object, nothing else is connected to it.